keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,660 @@
+side-effect free). Any parsing or type checking errors will cause the CRD write
+to fail with a descriptive error.
+
+#### Function library
+
+The function library available to expressions can be augmented using [extension
+functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#extension-functions).
+
+List of functions to include for the initial release:
+- Equality and Ordering (customized to do Kuberenetes semantic equality that handles "associative lists")
+- Regular Expressions
+- Some Standard Definitions
+
+
+Considerations:
+- The functions will become VERY difficult to change as this feature matures. We
+  should limit ourselves initially to functions that we have a high level of
+  confidence will not need to be changed or rethought.
+
+- Support kubernetes specific concepts, like accessing associative lists by key may be needed, but
+  we need to review more use cases to determine if this is needed.
+  
+- The Kubernetes associated list equality uses map semantic equality which is different from CEL. 
+  We would consider overwriting in CEL or adding a workaround utility function.
+
+
+### User Stories
+
+- Use case: [Tekton pipeline validation](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/pipeline_validation.go)
+  - Referential integrity checks
+  - Custom formatted validation error messages
+  - "Either list X or list Y must be non-empty"
+  - "There exists a X in one list/map for each Y in another map/list"
+- (PRs to add additional user stories to this list welcome!)
+
+### Notes/Constraints/Caveats (Optional)
+
+While we believe the expressiveness of CEL is pretty complete for our purposes, it is non-turing
+complete, and lacks support recursive data types (OpenAPIv3 & CRD structural schemas are not
+possible to validate with CEL).
+
+### Risks and Mitigations
+
+#### Accidental misuse
+
+Break the control plane by consuming excessive CPU and/or memory the api-server.
+
+Mitigation: CEL is specifically designed to constrain the running time of expressions and to limit
+the memory utilization. Since CRD creation is a privileged operation, it should be safe to
+integrate.
+
+#### Malicious use
+
+Breaking out of the sandbox to run untrusted code in the apiserver or exfiltrate data.
+
+Mitigation: CEL is designed to sandbox code execution. 
+
+
+### Future Plan
+  
+#### Defaulting
+
+The `x-kubernetes-default` extension will be used. The location of the defaulter
+determines the scope of defaulter.
+
+The 'field' specifies which field is to be defaulted using a field path that is
+relative to the scope root. The defaulter will only be run if 'field' is unset, 
+and the result of the expression must match the type of the field to be
+defaulted.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          x-kubernetes-default:
```

Comment:
is this a good idea? defaulting one field based on another has led to tears in the past and is not recommended

Reply:
Maybe not? I'll drop it for now. I'll trim up the future work section to focus on CEL for admission control and CRD conversion.